test/langtools/tools/javac/classreader/ParameterModifiersAcrossCompilationBoundaries.java

```diff
@@ -0,0 +1,156 @@
+                    }
+
+                    @Override
+                    public void init(JavacTask javacTask, String... strings) {
+                        javacTask.addTaskListener(
+                                new TaskListener() {
+                                    @Override
+                                    public void finished(TaskEvent e) {
+                                        if (e.getKind() != TaskEvent.Kind.ENTER) {
+                                            return;
+                                        }
+                                        Elements elements = javacTask.getElements();
+                                        TypeElement typeElement = elements.getTypeElement("B");
+                                        for (Element m : typeElement.getEnclosedElements()) {
+                                            if (m instanceof ExecutableElement) {
+                                                for (VariableElement p : ((ExecutableElement) m).getParameters()) {
+                                                    System.err.println(p.getSimpleName() + " " + p.getModifiers());
+                                                }
+                                            }
+                                        }
+                                    }
+                                });
+                    }
+                }
+                """;
+
+    String aCode = """
+            public class A { }
+            """;
+
+    String bCode = """
+            public class B {
+                void f(final int x) {
+                }
+            }
+            """;
+
+    public ParameterModifiersAcrossCompilationBoundaries() {
+        super(System.err);
+        tb = new ToolBox();
+    }
+
+    public static void main(String[] args) throws Exception {
+        var t = new ParameterModifiersAcrossCompilationBoundaries();
+        t.runTests();
+    }
+
+    @Test
+    public void testParameterModifiersNotVisible() throws Exception {
+        Path base = Path.of(".");
+        Path pluginPath = base.resolve("plugin");
+        Path testPath = base.resolve("test");
+        tb.writeFile(pluginPath.resolve("p").resolve("P.java"), pluginCode);
+        tb.writeFile(pluginPath.resolve("module-info.java"), moduleCode);
+        tb.writeFile(testPath.resolve("A.java"), aCode);
+        tb.writeFile(testPath.resolve("B.java"), bCode);
+
+        new JavacTask(tb)
+                .files("plugin/module-info.java", "plugin/p/P.java")
+                .run()
+                .writeAll();
+
+        List<String> firstOutput = new JavacTask(tb)
+                .options("--processor-module-path", "plugin", "-Xplugin:P",
+                         "-parameters")
+                .files("test/A.java", "test/B.java")
+                .run()
+                .writeAll()
+                .getOutputLines(Task.OutputKind.STDERR);
+        List<String> firstExpected = Arrays.asList("x [final]", "x [final]");
```

Comment:
Why are there supposed to be two "x [final]" in the first output?

Reply:
Because it has two source files. Please see the code snippet of the method `JavaCompiler.enterTrees`.

```
public List<JCCompilationUnit> enterTrees(List<JCCompilationUnit> roots)
// ignore other code.
        if (!taskListener.isEmpty()) {
            for (JCCompilationUnit unit: roots) {
                TaskEvent e = new TaskEvent(TaskEvent.Kind.ENTER, unit);
                taskListener.finished(e);
            }
        }
//  ignore other code.
}
```

The task listener is called every time there is a `JCCompilationUnit`.